libvirt/data_source_libvirt_node_device_info.go

```diff
@@ -0,0 +1,183 @@
+	Domain     int        `xml:"domain"`
+	Bus        int        `xml:"bus"`
+	Slot       int        `xml:"slot"`
+	Function   int        `xml:"function"`
+	Product    Product    `xml:"product"`
+	Vendor     Vendor     `xml:"vendor"`
+	IommuGroup IommuGroup `xml:"iommuGroup"`
+	// STORAGE
+	Block            string `xml:"block"`
+	DriveType        string `xml:"drive_type"`
+	Model            string `xml:"model"`
+	Serial           string `xml:"serial"`
+	Size             int    `xml:"size"`
+	LogicalBlockSize int    `xml:"logical_block_size"`
+	NumBlocks        int    `xml:"num_blocks"`
+	// USB
+	Number   int `xml:"number"`
+	Subclass int `xml:"subclass"`
+	Protocol int `xml:"protocol"`
+}
+
+type Product struct {
+	Name string `xml:",chardata"`
+	Id   string `xml:"id,attr"`
+}
+
+type Vendor struct {
+	Name string `xml:",chardata"`
+	Id   string `xml:"id,attr"`
+}
+
+type IommuGroup struct {
+	IommuGroup xml.Name  `xml:"iommuGroup"`
+	Number     int       `xml:"number,attr"`
+	Addresses  []Address `xml:"address"`
+}
+
+type Address struct {
+	Address  xml.Name `xml:"address"`
+	Domain   string   `xml:"domain,attr"`
+	Bus      string   `xml:"bus,attr"`
+	Slot     string   `xml:"slot,attr"`
+	Function string   `xml:"function,attr"`
+}
+
+// a libvirt nodeinfo datasource
+//
+// Datasource example:
+//
+//	data "libvirt_node_device_info" "info" {
+//	}
+//
+//	output "xml" {
+//	  value = data.libvirt_node_device_info.info.xml
+//	}
+func datasourceLibvirtNodeDeviceInfo() *schema.Resource {
+	return &schema.Resource{
+		Read: resourceLibvirtNodeDeviceInfoRead,
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"path": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"parent": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"xml": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"capability": {
+				Type:     schema.TypeMap,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceLibvirtNodeDeviceInfoRead(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] Read data source libvirt_nodedevices")
+
+	virConn := meta.(*Client).libvirt
+	if virConn == nil {
+		return fmt.Errorf(LibVirtConIsNil)
+	}
+
+	var device_name string
+
+	if name, ok := d.GetOk("name"); ok {
+		device_name = name.(string)
+		log.Printf("[DEBUG] Got name: %s", device_name)
+	}
+
+	device, err := virConn.NodeDeviceLookupByName(device_name)
+	if err != nil {
+		return fmt.Errorf("failed to lookup node device: %v", err)
+	}
+
+	xml_desc, err := virConn.NodeDeviceGetXMLDesc(device.Name, 0)
+	if err != nil {
+		return fmt.Errorf("failed to get XML for node device: %v", err)
+	}
+
+	var device_xml Device
+
+	err = xml.Unmarshal([]byte(xml_desc), &device_xml)
+	if err != nil {
+		log.Fatalf("failed to unmarshal device_xml into XML: %v", err)
+	}
+
+	capability := map[string]interface{}{}
+	if device_xml.Capability.Type == "pci" {
```

Comment:
Any reason we don't support [all the `Type` options](https://libvirt.org/formatnode.html#capability-types) from the get go? Seems fine to only support a select few at first, just curious.

Reply:
Added the rest of relevant devices, with the exception of S390 hardware which, if someone has access to, can add.